### PR TITLE
#3391 multiViewState for dataList V2

### PR DESCRIPTION
**Pull request by christophs78**
I screwed my existing PR during rebase. Now the clean PR again.

### CODE REVIEW

File: src/main/java/org/primefaces/component/datalist/DataList.java [hunk 1]
```diff
@@ -195,4 +196,42 @@ protected void processChildren(FacesContext context, PhaseId phaseId) {
         }
     }
 
+    @Override
+    public boolean isMultiViewState() {
```
Comment (Rapster) [inline src/main/java/org/primefaces/component/datalist/DataList.java]: This method has no use. Please delete it

File: src/main/java/org/primefaces/component/datalist/DataList.java [hunk 2]
```diff
@@ -195,4 +196,42 @@ protected void processChildren(FacesContext context, PhaseId phaseId) {
...
+        return super.isMultiViewState();
+    }
+
+    public void restoreListState() {
+        ListState ls = getListState(false);
+        if (ls != null) {
+            if (isPaginator()) {
```
Comment (Rapster) [inline src/main/java/org/primefaces/component/datalist/DataList.java]: Make one condition using "&&"

File: src/main/java/org/primefaces/component/datalist/DataList.java [hunk 3]
```diff
@@ -195,4 +196,42 @@ protected void processChildren(FacesContext context, PhaseId phaseId) {
...
+        }
+    }
+
+    public ListState getListState(boolean create) {
+        FacesContext fc = getFacesContext();
+        Map<String, Object> sessionMap = fc.getExternalContext().getSessionMap();
+        Map<String, ListState> dlState = (Map) sessionMap.get(Constants.DATALIST_STATE);
+        String viewId = fc.getViewRoot().getViewId().replaceFirst("^/*", "");
+        String stateKey = viewId + "_" + getClientId(fc);
+        ListState ls;
```
Comment (Rapster) [inline src/main/java/org/primefaces/component/datalist/DataList.java]: Write `ListState ls = dlState.get(stateKey);` instead, before the condition (e.g remove that cumbersome declaration `ListState ls;`)

**Review by Rapster — COMMENTED** [review]
@christophs78 I started a review, please see the comments

**Comment by christophs78** [comment]
@Rapster Thank you for reviewing! Good point´s! Also made one modification to the existing DataTable-class which had the same issue.

**Comment by tandraschko** [comment]
i would call the ListState class DataListState

**Comment by christophs78** [comment]
Ok, I renamed ListState to DataListState.
Originally named it ListState because DataTable stores it´s multiViewState in org.primefaces.component.datatable.TableState. Tried to do replicate this naming for multiViewState of DataList.

**Comment by tandraschko** [comment]
Yep, i know your intensions but even TableState is to generic as we also have a TreeTable.

**Comment by tandraschko** [comment]
Thanks for your (great) first PR ;)

**Pull request merged**